launch/launch/substitutions/equals_substitution.py

```diff
@@ -0,0 +1,123 @@
+def _str_is_bool(input: Text) -> bool:
+    """Check if string input is convertible to a boolean."""
+    if not isinstance(input, Text):
+        return False
+    else:
+        return input.lower() in ('true', 'false', '1', '0')
+
+
+def _str_is_float(input: Text) -> bool:
+    """Check if string input is convertible to a float."""
+    try:
+        float(input)
+        return True
+    except ValueError:
+        return False
+
+
+@expose_substitution('equals')
+class EqualsSubstitution(Substitution):
+    """
+    Substitution that checks if two inputs are equal.
+
+    Returns 'true' or 'false' strings depending on the result.
+    """
+
+    def __init__(
+        self,
+        left: Optional[Union[Any, Iterable[Any]]],
+        right: Optional[Union[Any, Iterable[Any]]]
+    ) -> None:
+        """Create an EqualsSubstitution substitution."""
+        super().__init__()
+
+        if is_some_substitutions_type(left):
+            self.__left = normalize_to_list_of_substitutions(left)
+        elif left is None:
+            self.__left = ""
+        elif isinstance(left, bool):
+            self.__left = str(left).lower()
+        else:
+            self.__left = str(left)
+
+        if is_some_substitutions_type(right):
+            self.__right = normalize_to_list_of_substitutions(right)
+        elif right is None:
+            self.__right = ""
+        elif isinstance(right, bool):
+            self.__right = str(right).lower()
+        else:
+            self.__right = str(right)
+
+    @classmethod
+    def parse(cls, data: Iterable[SomeSubstitutionsType]):
+        """Parse `EqualsSubstitution` substitution."""
+        if len(data) != 2:
+            raise TypeError('and substitution expects 2 arguments')
+        return cls, {'left': data[0], 'right': data[1]}
+
+    @property
+    def left(self) -> Substitution:
+        """Getter for left."""
+        return self.__left
+
+    @property
+    def right(self) -> Substitution:
+        """Getter for right."""
+        return self.__right
+
+    def describe(self) -> Text:
+        """Return a description of this substitution as a string."""
+        return f'EqualsSubstitution({self.left} {self.right})'
+
+    def perform(self, context: LaunchContext) -> Text:
+        """Perform the substitution."""
+        if is_some_substitutions_type(self.left) and not isinstance(self.left, Text):
+            left = perform_substitutions(context, self.left)
+        else:
+            left = self.left
+
+        if is_some_substitutions_type(self.right) and not isinstance(self.right, Text):
+            right = perform_substitutions(context, self.right)
+        else:
+            right = self.right
+
+        if _str_is_bool(left) and _str_is_bool(right):
+            return str((left.lower() in ('true', '1')) == (right.lower() in ('true', '1'))).lower()
+        elif _str_is_float(left) and _str_is_float(right):
+            return str(math.isclose(float(left), float(right))).lower()
```

Comment:
I'm not convinced we should add these special cases. It seems like it would be less confusing if we just clearly document that this substitution is just doing a string comparison (and that any provided arguments are cast to strings). I'm on the fence about it, but whatever we decide I think it needs to be clearly documented (ie. what does it mean for two arguments to be equal?).

Reply:
I'm thinking of the cases (well, at least in ROS 1...) where I've seen people put in `'1'` instead of `'true'` that motivated the addition of those special cases.

Likewise for the Python launch API, the naming of the substitutions remind me of the Python constructs, which make me want to have the substitutions mimic that behavior.

I'm not sure what the decision should be on this though. I guess it depends on whether we think more people will use xml vs Python, and what people will find intuitive...

Till then, I think I'll add a note for these cases in the docstring and continue supporting them? :thinking:

Reply:
Looking around the existing code, I see we define valid "true" and "false" values here:

https://github.com/ros2/launch/blob/76ade26d3f7bcb6a5a1b72a512a8c098e6d53f04/launch/launch/conditions/evaluate_condition_expression_impl.py#L25-L26

So, it's okay to support the string `'1'` and, _maybe_, the number `1`.

For comparison, `IfCondition` and `UnlessCondition` (and I presume other conditions) expect a string. For example, the following launch file results in an exception:

```yaml
launch:
- group:
    if: 1
    children:
        - log:
            message: Hello
```

I would clearly document the expectations of the current implementation and I think that would be enough unless there's a dissenting opinion.

Reply:
[This PR](https://github.com/ros2/launch/pull/651) should adopt '1' and '0' truthiness for the And, Or, and Not conditions.

Once that's merged in I will make use of the type coercion for the Any and All conditions

Reply:
Implemented and merged in https://github.com/ros2/launch/pull/651

Reply:
Correction, I couldn't use the implementation in #651 here because the Equals substitution cannot hard-enforce bool string inputs.

So the string checking I'm doing here persists. (Checking for 'True', 'true' and '1', and the equivalent for false only when both inputs are like that.)